cmd/tsdb/main.go

```diff
@@ -188,57 +329,110 @@ func (b *writeBenchmark) ingestScrapes(lbls []labels.Labels, scrapeCount int) (u
 	return total, nil
 }
 
-func (b *writeBenchmark) ingestScrapesShard(metrics []labels.Labels, scrapeCount int, baset int64) (uint64, error) {
+func (b *writeBenchmark) ingestScrapesShard(metrics []labels.Labels, m3dbMetrics []ident.TagsIterator, scrapeCount int, baset int64) (uint64, error) {
 	ts := baset
-
-	type sample struct {
-		labels labels.Labels
-		value  int64
-		ref    *uint64
-	}
-
-	scrape := make([]*sample, 0, len(metrics))
-
-	for _, m := range metrics {
-		scrape = append(scrape, &sample{
-			labels: m,
-			value:  123456789,
-		})
-	}
 	total := uint64(0)
 
-	for i := 0; i < scrapeCount; i++ {
-		app := b.storage.Appender()
-		ts += timeDelta
+	if b.useM3DB() {
+		type sample struct {
+			tags  ident.TagsIterator
+			value int64
+			ref   *uint64
+		}
 
-		for _, s := range scrape {
-			s.value += 1000
+		scrape := make([]*sample, 0, len(metrics))
 
-			if s.ref == nil {
-				ref, err := app.Add(s.labels, ts, float64(s.value))
-				if err != nil {
-					panic(err)
-				}
-				s.ref = &ref
-			} else if err := app.AddFast(*s.ref, ts, float64(s.value)); err != nil {
+		for _, m := range m3dbMetrics {
+			scrape = append(scrape, &sample{
+				tags:  m,
+				value: 123456789,
+			})
+		}
 
-				if errors.Cause(err) != tsdb.ErrNotFound {
-					panic(err)
+		batch := make(commitlog.WritesBatch, len(scrape))
+		for i := 0; i < scrapeCount; i++ {
+			ts += timeDelta
+			id := ident.StringID(string(i))
+			for j, s := range scrape {
+				s.value += 1000
+				batch[j] = commitlog.Write{
+					Series: commitlog.Series{
+						Namespace: defaultM3DBNamespace,
+						ID:        id,
```

Comment:
`s.tags` obivously differs for each here but `ID` stays the same – what does that mean?

Do I read correctly, that all samples of a scrape are appended to the same series with an ID of simply the scrape index?
The tags would obviously be different so that wouldn't seem right. Using `j` would not be correct either since scrapes are not guaranteed to always contain the same series in the same order.

That looks like it is always doing an equivalent of `AddFast`. The main problem in TSDB is really getting from an `s.tags` to an`ID`. That's why `Add` is so much slower (each insert essentially requires a full index lookup) and there's dedicated `AddFast` path.

So I guess my question is whether I misunderstand what `ID` stands for or how we would be able to figure out the correct one if all we have is a set of tags.

Reply:
Ah ok I may have done the wrong thing here. I'm still new to Prometheus / TSDB so still trying to figure out the model.

Is the way this is supposed to work is that each individual "scrape" is a different "series"? If so, I can fix the benchmark. The confusion lies around the difference in-between how TSDB and M3DB identify a given series. It seems like in TSDB a series is identified by a combination of tags, whereas in M3DB every series has a unique ID and (optionally) a bunch of tags attached to that ID.

If you write different tags to the same ID the new tags will just be ignored. 

I don't think we're doing the equivalent of AddFast because even with the WriteTaggedBatch API we still have to look up each series in-memory for each individual write. The WriteTaggedBatch is basically just an optimization to avoid a lot of contention on our commitlog queue, but doesn't do any lookup optimization.

In terms of figuring out the correct ID given only a set of tags (for the write path at least) I think we'd need to do something that deterministically generates the same ID from a given set of tags, I think thats what our M3Coordinator service does when its receiving data from Prometheus over the remote read/write endpoint and then inside of M3DB we have some optimizations to re-use the memory that is allocated for the tags within the generated ID so it has minimal impact on memory utilization

Reply:
Each sample in a scrape belongs to a different series in the general case – at least in the way Prometheus does writes.
When we scrape a `/metrics` endpoint, a bunch of tags is all we get to identify what a sample belongs to. The targets have no way to tell us at scrape time what database ID a series they expose map to.
So figuring out what series that has to be written to is the fundamental difficulty and makes up most of the work TSDB does.

A simple way for m3db would probably be to use a sorted and serialized version of the tags/labels as the ID. This is usually rather expensive, which is why Prometheus+TSDB play quite a few tricks with Add and AddFast.